intel_extension_for_transformers/llm/runtime/graph/core/shared_memory_ccl.hpp

```diff
@@ -0,0 +1,222 @@
+//  Copyright (c) 2023 Intel Corporation
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+#include <assert.h>
+#include <fcntl.h>
+#include <immintrin.h>
+#include <omp.h>
+#include <sys/mman.h>
+#include <unistd.h>
+// #include <atomic>
+#include "oneapi/ccl.hpp"
+
+// states for collectives
+enum ccl_state {
+  ccl_begin = 0,
+  copy_in_done,
+  reduce_done,
+  copy_out_done,
+};
+
+void* shared_open(const char* name, size_t nbytes) {
+  int d = shm_open(name, O_RDWR, S_IRUSR | S_IWUSR);
+  if (d != -1) {
+    return mmap(NULL, nbytes, PROT_READ | PROT_WRITE, MAP_SHARED, d, 0);
+  } else {
+    printf("shared_open %s failed\n", name);
+    return nullptr;
+  }
+}
+
+void* shared_create(const char* name, void* bytes, size_t nbytes) {
+  int d = shm_open(name, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
+  if ((d != -1) && (nbytes = write(d, bytes, nbytes))) {
+    return mmap(NULL, nbytes, PROT_READ | PROT_WRITE, MAP_SHARED, d, 0);
+  } else {
+    printf("shared_create %s failed\n", name);
+    return nullptr;
+  }
+}
+
+void shared_close(const char* name, void* bytes, size_t nbytes) {
+  int d = shm_open(name, O_RDWR, S_IRUSR | S_IWUSR);
+  if (d != -1) {
+    munmap(bytes, nbytes);
+    shm_unlink(name);
+  }
+}
+
+#define MAX_BUF_SIZE 1048576
+struct ccl_buffer {
+  enum ccl_state state;
+  char data[MAX_BUF_SIZE];
+};
```

Comment:
OK, it will actually ensure data is algen to 4-byts (enum => int => 4byts). But anyway, I suggest aligning `data` to 32(avx2) / 64(avx512) bytes for better performance by avoiding inter-cacheline access.